docs/setup/forward_proxy.md

```diff
@@ -0,0 +1,71 @@
+# Using a forward proxy with Synapse
+
+You can use Synapse with a forward or outbound proxy. This will be neccssary
+for expamle in corporate environments behind a DMZ (demilitarized zone).
+Synapse supports routing outbound HTTP(S) requests via a proxy. Only HTTP(S)
+proxy is supported, not SOCKS proxy or anything else.
+
+## Configure
+
+The `http_proxy`, `https_proxy`, `no_proxy` environment variable is used to
+specify proxy settings. The environment variable is not case sensitive.
+- `http_proxy`: Proxy server to use for HTTP requests.
+- `https_proxy`: Proxy server to use for HTTPS requests.
+- `no_proxy`: Comma-separated list of hosts, IP addresses, or IP ranges in CIDR
+  format which do not use a proxy and Synapse use a direct connection.
+
+The proxy setting has the form: `[scheme://][<username>:<password>@]<host>[:<port>]`
+- Supported scheme is `http://` and `https://`. Default scheme is `http://`.
+  It is recommended to set a scheme. The default value is for compatibility
+  reasons. If scheme is set to `https://` the connection uses TLS between
+  Synapse and proxy.
+
+  **NOTE**: Synapse validates the certificates. If the certificate is not
+  valid, then the connection is dropped.
+- Default port if not given is `1080`.
+- Username and password is optional and will be used for authenticate against
+  proxy.
+
+**Examples**
+- HTTP_PROXY=http://USERNAME:PASSWORD@10.0.1.1:8080/
+- HTTPS_PROXY=http://USERNAME:PASSWORD@proxy.example.com:8080/
+- NO_PROXY=master.hostname.example.com,10.1.0.0/16,172.30.0.0/16
+
+**NOTE**:
+Blacklisting IP addresses does not work for connections throug a proxy because
+the proxy server resolves DNS names and not Synapse. In this case you have to
+use your firewall or proxy for blacklisting IP addresses.
+
+## Connection types
+
+The proxy will be **used** for:
+
+- push
+- url previews
+- phone-home stats
+- recaptcha validation
+- CAS auth validation
```

Comment:
Is this list from someplace we already have or did you look through the code?

This seems to match my memory, but curious if we should double check!

Reply:
The list is from there #6239

Reply:
We updated this at some point, I think https://github.com/matrix-org/synapse/blob/016f085722666ac68d2e4b4f21793db399203ea9/docs/sample_config.yaml#L205-L206 is the most up-to-date information.

Reply:
I am not sure if I understand you.
The documentation is from Dec 2020. There was a change in Jan 2021 #9084
Does the `config.sample` needs a info for do not block if a proxy is used?

Reply:
My point was that the comment from #6239 didn't take into account the changes made in #9084 so I think the current list is wrong. (It says that we don't use it for federation and identity servers, but we do now).

Reply:
This really is not trivial.
I took a look in code, now:
- proxy
https://github.com/matrix-org/synapse/blob/bf72d10dbf506f5ea486d67094b6003947d38fb7/synapse/server.py#L408
  - phone_stats_home
  - cas
  - oidc
  - ui_auth/checkers
https://github.com/matrix-org/synapse/blob/bf72d10dbf506f5ea486d67094b6003947d38fb7/synapse/server.py#L415
  - handlers/federation
    - "Checks whether public_key has been revoked"
  - httppusher
  - `class PreviewUrlResource`
- no proxy
https://github.com/matrix-org/synapse/blob/bf72d10dbf506f5ea486d67094b6003947d38fb7/synapse/server.py#L428
  - keyring
  - federation/transport/client
    - "Sends federation HTTP requests to other servers"
  - identity
  - media_repository
https://github.com/matrix-org/synapse/blob/bf72d10dbf506f5ea486d67094b6003947d38fb7/synapse/server.py#L401
  - generic_worker
  - module_api
  - replication
  - `class IdentityHandler`

Reply:
Oh I'm being an idiot -- I was confusing the proxy with the IP blacklisting code, which of course aren't the same. 😢

Reply:
Ok. This may happen.
But I am confused about the blacklist.
How can Synapse blacklist when the proxies does the DNS resolution? Synapse does not know the IP.
In this PR #10129 a warning should be added. But there is a function in code like:
https://github.com/matrix-org/synapse/blob/bf72d10dbf506f5ea486d67094b6003947d38fb7/synapse/server.py#L415-L418

Reply:
> How can Synapse blacklist when the proxies does the DNS resolution?

It doesn't, it only applies the blacklist if a proxy isn't being used, so `get_proxied_blacklisted_http_client` does the following:

* Checks if a proxy is used, if so uses it.
* Otherwise, doesn't use a proxy and applies the blacklist.

Pretty much anything that uses `get_proxied_blacklisted_http_client` or `get_proxied_http_client` uses the proxy.

Frankly we should probably change anywhere that manually creates `SimpleHttpClient` to call one of the above (that's part of fixing "make proxies work everywhere IMO).